### ISSUE TITLE
Question regards the loss

### ISSUE THREAD

**Issue by abduallahmohamed** [tonyduan/mixture-density-network#1]
Hi, 

in
https://github.com/tonyduan/mdn/blob/master/mdn/models.py#L29 
```

        pi, normal = self.forward(x)
        loglik = normal.log_prob(y.unsqueeze(1).expand_as(normal.loc))
        loglik = torch.sum(loglik, dim=2)
        loss = -torch.logsumexp(torch.log(pi.probs) + loglik, dim=1)
```

Should be: 

```
        pi, normal = self.forward(x)
        loglik = normal.log_prob(y.unsqueeze(1).expand_as(normal.loc))
        loglik +=torch.log(pi.probs)
        loss = -torch.mean( loglik)

```

Or I'm mistaken about it?

**Comment by tonyduan** [comment]
I think it's correct as it is. What you propose is a mixture of log-probabilities, not the log-probability of a mixture model. More concretely, the log-likelihood we want can be expressed:

$$\log\sum_k \pi_k \mathcal{N}(x;\mu_k,\sigma_k) = \log \sum_k \exp(\log \pi_k + \log \mathcal{N}(x;\mu_k,\sigma_k))$$

What you have is instead:

$$\sum_k (\log\pi_k + \log\mathcal{N}(x;\mu_k,\sigma_k))$$
